### PR TITLE
Support building sqlite / room on Github

### PR DESCRIPTION
**Pull request by yigit**
This CL adds sqlite as a project to GitHub and makes it possible compile.

On AOSP, we use sqlite prebuilts to find sources. On GitHub, we'll download sqlite
amalgamation. To ensure they stay consistent, we check the version in the header
file which will also be checked on AOSP build.

Fixes: 306669673
Test: ./gradlew playgroundCIHostTests

### CODE REVIEW

File: buildSrc/private/src/main/kotlin/androidx/build/PrepareSqliteSourcesTask.kt [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+ * [destinationDirectory].
+ * On Github, they are downloaded from SQLite servers.
+ *
+ * To ensure each version is consistent, we use the [sqliteVersion] parameter and check the SQLite
+ * source code for them.
+ */
+@DisableCachingByDefault(
+    because = "This is a simple copy/validation task that is not worth caching"
+)
+abstract class PrepareSqliteSourcesTask @Inject constructor(
```
Comment (yigit) [inline buildSrc/private/src/main/kotlin/androidx/build/PrepareSqliteSourcesTask.kt]: i'm not sure if this should live in public or private buildSrc.
it is only used by the sqlite-bundled task so we could consider moving this into that build.gradle file but the logic is not straightforward, hence moved into buildSrc.

File: buildSrc/private/src/main/kotlin/androidx/build/KonanPrebuiltsSetup.kt [hunk 2]
```diff
@@ -34,16 +34,21 @@ object KonanPrebuiltsSetup {
 
     /**
      * Creates a Konan distribution with the given [prebuiltsDirectory] and [konanHome].
+     *
+     * @param prebuiltsDirectory The directory where AndroidX prebuilts are present. Can be `null`
```
Comment (liutikas) [inline buildSrc/private/src/main/kotlin/androidx/build/KonanPrebuiltsSetup.kt:38]: make a note that null means fetching konan from network.

File: buildSrc/private/src/main/kotlin/androidx/build/PrepareSqliteSourcesTask.kt (hunk 1 above)
Comment (liutikas) [reply buildSrc/private/src/main/kotlin/androidx/build/PrepareSqliteSourcesTask.kt]: this is super sqlite specific and not even used in aosp, i'd say we move this to sqlite build.gradle

File: sqlite/sqlite-bundled/build.gradle [hunk 3]
```diff
@@ -76,6 +82,14 @@ configurations {
             attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, Usage.NATIVE_LINK))
         }
     }
+    // Configuration for producing a shareable directory that includes SQLite amalgamation sources.
+    sqliteSources {
```
Comment (liutikas) [inline sqlite/sqlite-bundled/build.gradle:90]: side question: what do we want to do with sources of sqlite that be bundle? should they ship in source jars?

File: buildSrc/private/src/main/kotlin/androidx/build/PrepareSqliteSourcesTask.kt (hunk 1 above)
Comment (yigit) [reply buildSrc/private/src/main/kotlin/androidx/build/PrepareSqliteSourcesTask.kt]: done, moved to a groovy file.

File: sqlite/sqlite-bundled/build.gradle (hunk 3 above)
Comment (yigit) [reply sqlite/sqlite-bundled/build.gradle:90]: @danysantiago , wdyt? 
imo, might be nice for completeness but the only API seen by consumer is the sqlite driver impl created by us (they don't get direct access to sqlite though it might be possible to load it maybe).

File: buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt [hunk 4]
```diff
@@ -233,9 +236,11 @@ abstract class KonanBuildService @Inject constructor(
                 it.parameters.konanHome.set(
                     nativeCompilerDownloader.compilerDirectory
                 )
-                it.parameters.prebuilts.set(
-                    project.getKonanPrebuiltsFolder()
-                )
+                if (!ProjectLayoutType.isPlayground(project)) {
+                    it.parameters.prebuilts.set(
+                        project.getKonanPrebuiltsFolder()
```
Comment (liutikas) [inline buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt:267]: Can we assert it is not null here? I would not want to accidentally start downloading these in CI.

File: buildSrc/private/src/main/kotlin/androidx/build/clang/NativeTargetCompilation.kt [hunk 5]
```diff
@@ -137,4 +140,29 @@ class NativeTargetCompilation internal constructor(
             }
         }
     }
+
+    /**
+     * JDK ships with JNI headers only for the current platform. As a result,we don't have access
```
Comment (liutikas) [inline buildSrc/private/src/main/kotlin/androidx/build/clang/NativeTargetCompilation.kt]: missing space after `,`

File: buildSrc/private/src/main/kotlin/androidx/build/clang/NativeTargetCompilation.kt [hunk 6]
```diff
@@ -137,4 +140,29 @@ class NativeTargetCompilation internal constructor(
...
+     * to cross-platform jni headers. They are mostly the same and we don't ship cross compiled code
+     * from GitHub so it is acceptable to use local JNI headers for cross platform compilation on
+     * GitHub.
+     */
+    private fun findJniHeadersInPlayground(
+        javaHome: File
+    ): List<File> {
+        var include = File(javaHome, "include")
+        if (!include.exists()) {
```
Comment (liutikas) [inline buildSrc/private/src/main/kotlin/androidx/build/clang/NativeTargetCompilation.kt]: why is this needed? Please leave a comment

File: sqlite/sqlite-bundled/build.gradle [hunk 7]
```diff
@@ -55,19 +58,23 @@ def SQLITE_COMPILE_FLAGS = [
...
-// nice to have them there so that we can swap them with SQLite amalgamation for Github.
-// TODO(b/306669673): Support GitHub builds
-def copySqliteSourcesTask = tasks.register("copySqliteSources", Copy) {
-    from(new File(AndroidXConfig.getExternalProjectPath(project), "sqlite/dist/orig")) {
-        include "sqlite3.c"
-        include "sqlite3.h"
+// see PrepareSqliteSourcesTask.groovy for implementation details of the register method
+def prepareSqliteSourcesTask = registerPrepareSqliteSourcesTask(
+        project, // project
+        "prepareSqliteSourcesTask" // task name
```
Comment (liutikas) [inline sqlite/sqlite-bundled/build.gradle:67]: you can use named arguments in groovy
registerPrepareSqliteSourcesTask(project: project, name: "prepareSqliteSourcesTask")

File: sqlite/sqlite-bundled/build.gradle [hunk 8]
```diff
@@ -55,19 +58,23 @@ def SQLITE_COMPILE_FLAGS = [
...
 
-// Copy SQLite sources into a build directory. Even though this is not needed, it will be
-// nice to have them there so that we can swap them with SQLite amalgamation for Github.
-// TODO(b/306669673): Support GitHub builds
-def copySqliteSourcesTask = tasks.register("copySqliteSources", Copy) {
-    from(new File(AndroidXConfig.getExternalProjectPath(project), "sqlite/dist/orig")) {
-        include "sqlite3.c"
-        include "sqlite3.h"
+// see PrepareSqliteSourcesTask.groovy for implementation details of the register method
+def prepareSqliteSourcesTask = registerPrepareSqliteSourcesTask(
```
Comment (liutikas) [inline sqlite/sqlite-bundled/build.gradle:65]: somewhere within this file we should have a comment explaining the license of sqlite that we are relicensing it here as apache 2.0

**Review by liutikas — APPROVED** [review]

File: sqlite/sqlite-bundled/build.gradle (hunk 7 above)
Comment (yigit) [reply sqlite/sqlite-bundled/build.gradle:67]: not sure about that.
as far as i can see, it requires me to use a Map as an argument in the function declaration, which would make the function declaration less useful.

https://docs.groovy-lang.org/latest/html/documentation/#_named_parameters_2

am i missing something ?

File: buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt (hunk 4 above)
Comment (yigit) [reply buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt:267]: not sure if i understand.
this is where we set the parameters so there is nothing to assert.

I could try to assert when we create the distribution (where we read the parameter) but the build service doesn't have access to `Project` so I'm not sure if i can check for ` if(!playground) assertNotNullPrebuilts()`. one option is to pass the project layout type as an argument to the build service.

Is that what you prefer or am i misunderstanding this?

File: buildSrc/private/src/main/kotlin/androidx/build/clang/NativeTargetCompilation.kt (hunk 6 above)
Comment (yigit) [reply buildSrc/private/src/main/kotlin/androidx/build/clang/NativeTargetCompilation.kt]: this is from my kotlin-sqlite-bindings copy :). i don't remember why so let me try removing it.

File: sqlite/sqlite-bundled/build.gradle (hunk 8 above)
Comment (yigit) [reply sqlite/sqlite-bundled/build.gradle:65]: added a note on licensing.

File: buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt (hunk 4 above)
Comment (yigit) [reply buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt:267]: implemented the option where i pass project layout type.

Comment (liutikas) [reply buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt:267]: where is this implementation? I cannot seem to find it.

File: sqlite/sqlite-bundled/build.gradle (hunk 7 above)
Comment (liutikas) [reply sqlite/sqlite-bundled/build.gradle:67]: You are right, I misunderstood how groovy does it. we can leave it as-is

File: buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt (hunk 4 above)
Comment (yigit) [reply buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt:267]: mb. the code push didn't go through 🤦  updated now.

Comment (yigit) [reply buildSrc/private/src/main/kotlin/androidx/build/clang/KonanBuildService.kt:267]: FYI this is the check: https://github.com/androidx/androidx/pull/647/files#diff-4d57a5bf411f5c52d9b975b078f74b4c8034a31922c196ad836485a03f1f130fR56
